File: src/main/java/view25d/prims/SetObserverFocus.java
package view25d.prims;

import org.nlogo.api.Argument;
import org.nlogo.api.Context;
import org.nlogo.api.DefaultCommand;
import org.nlogo.api.ExtensionException;
import org.nlogo.api.LogoException;
import org.nlogo.api.Syntax;

import view25d.View25DExtension;
import view25d.view.MouseableGLWindow;
import view25d.view.VarviewWindow;

public class SetObserverFocus extends DefaultCommand {

	@Override
    public Syntax getSyntax() {
       int[] argTypes = {Syntax.StringType(), Syntax.NumberType(), Syntax.NumberType()};
       return  Syntax.commandSyntax(argTypes);
    }
	
	@Override
	public String getAgentClassString() {
		return "O";
	}
	
	@Override
	public void perform(Argument[] args, Context context)
			throws ExtensionException, LogoException {
		
		String whichView = args[0].getString();
		double newX = args[1].getDoubleValue();
		double newY = args[2].getDoubleValue();
		
		VarviewWindow vvwp = View25DExtension.patchWindowMap.get(whichView);
		
 		VarviewWindow vvwt = View25DExtension.turtleWindowMap.get(whichView);
 		
		if (vvwp != null) {
			MouseableGLWindow win = vvwp.getGLWindow();
			double [] coords = win.getObserverPerspectiveFocusPoint();
			
			double oldX = coords[0];
			double delX = newX - oldX;
			
			double oldY = coords[1];
			double delY = newY - oldY;

			win.shiftObserverFocusPoint(delX, delY);
		}
		
		if (vvwt != null) {
			MouseableGLWindow win = vvwt.getGLWindow();
			double [] coords = win.getObserverPerspectiveFocusPoint();
			
			double oldX = coords[0];
			double delX = newX - oldX;
			
			double oldY = coords[1];
			double delY = newY - oldY;

			win.shiftObserverFocusPoint(delX, delY);
		}

	}

}

File: src/main/scala/prims/SetObserverFocus.scala
package view25d.prims

import org.nlogo.api.{ Argument, Command, Context }
import org.nlogo.core.Syntax

import view25d.View25DExtension

class SetObserverFocus extends Command {
  override def getSyntax: Syntax =
    Syntax.commandSyntax(right = List(Syntax.StringType, Syntax.NumberType, Syntax.NumberType),
                         agentClassString = "O---")

  override def perform(args: Array[Argument], context: Context): Unit = {
    val title = args(0).getString
    val x = args(1).getDoubleValue
    val y = args(2).getDoubleValue

    View25DExtension.patchWindowMap.get(title).foreach(window => {
      val coords = window.getObserverPerspectiveFocusPoint

      window.shiftObserverFocusPoint(x - coords(0), y - coords(1))
    })

    View25DExtension.turtleWindowMap.get(title).foreach(window => {
      val coords = window.getObserverPerspectiveFocusPoint

      window.shiftObserverFocusPoint(x - coords(0), y - coords(1))
    })
  }
}
